opm/grid/cpgrid/CpGrid.cpp

```diff
@@ -467,129 +493,1199 @@ CpGrid::scatterGrid(EdgeWeightMethod method,
+                    leaf_cell_to_point[leafCellIdx][corn] = level_to_leaf_corners[{0, old_cell_to_point[corn]}];
+                }
+            }
+            // Cell to face.
+            for (const auto& face : old_cell_to_face)
+            {   // Auxiliary bool to identity boundary patch faces
+                bool is_there_face = false;
+                for(const auto& bound_face : boundary_patch_faces) {
+                    is_there_face = is_there_face || (face.index() == bound_face); //true-> coincides with one boundary patch face
+                    if (is_there_face)
+                        break;
+                }
+                if(is_there_face) { // Face belongs to the patch boundary.
+                    for (const auto& level_newFace : old_to_new_boundaryPatchFaces[{0, face.index()}]) {
+                        aux_cell_to_face[leafCellIdx].push_back({level_to_leaf_faces[level_newFace], face.orientation()});
+                    }
+                }
+                else { // Face does not belong to the patch boundary.
+                    aux_cell_to_face[leafCellIdx].push_back({level_to_leaf_faces[{0, face.index()}], face.orientation()});
+                }
+            }
+        }
+        else { // Refined cells. (Cell comes from level1)
+            // Get level where cell was created and its local index, to later deduce its parent.
+            auto& [l1, l1Idx]  = leaf_to_level_cells[leafCellIdx]; // {1, cell index in level 1}
+            leaf_child_to_parent_cells[leafCellIdx] = l1_child_to_parent_cells[l1Idx]; //(*data_[l1]).child_to_parent_[l1Idx];
+            // Cell to point.
+            for (int corn = 0; corn < 8; ++corn) {
+                leaf_cell_to_point[leafCellIdx][corn] = level_to_leaf_corners[{1, old_cell_to_point[corn]}];
+            }
+            // Cell to face.
+            for (auto& face : old_cell_to_face) {
+                aux_cell_to_face[leafCellIdx].push_back({level_to_leaf_faces[{1, face.index()}], face.orientation()});
+            }
+        }
+    }
+    // Leaf view cell to face.
+    for (int cell = 0; cell < cell_count; ++cell) {
+        leaf_cell_to_face.appendRow(aux_cell_to_face[cell].begin(), aux_cell_to_face[cell].end());
+    }
+    // Leaf view face to cell.
+    leaf_cell_to_face.makeInverseRelation(leaf_face_to_cell);
+    //  Add Leaf View to data_.
+    (this-> data_).push_back(leaf_view_ptr);
+    current_view_data_ = data_[2].get();
+}
+
 
 } // namespace Dune
+
+#define OPM_INSTANTIATE_PARTITION_ITERATE(method, Method)               \
+    template CpGridTraits::template Codim<0>:: Method CpGrid:: method<0>() const; \
+    template CpGridTraits::template Codim<1>:: Method CpGrid:: method<1>() const; \
+    template CpGridTraits::template Codim<3>:: Method CpGrid:: method<3>() const; \
+    OPM_INSTANTIATE_PARTITION_ITERATE(lbegin, LevelIterator);           \
+    OPM_INSTANTIATE_PARTITION_ITERATE(lend, LevelIterator);
```

Comment:
This macro, as it's placed at the end of the file, appears to be unused.  Do we need it?

Reply:
> It's needed for `CpGrid::lbegin()` and `CpGrid::lend()` (the level iterators).

Right, but as written there were no calls to the macro, and since the macro was at the end of the CPP file that means it was not used at all.  The backslash at the end of the `Codim<3>` would fuse the next line into the macro definition.  I **think** you may perhaps have wanted to write
```C++
#define OPM_INSTANTIATE_PARTITION_ITERATE(method, Method)               \
    template CpGridTraits::template Codim<0>:: Method CpGrid:: method<0>() const; \
    template CpGridTraits::template Codim<1>:: Method CpGrid:: method<1>() const; \
    template CpGridTraits::template Codim<3>:: Method CpGrid:: method<3>() const

OPM_INSTANTIATE_PARTITION_ITERATE(lbegin, LevelIterator);
OPM_INSTANTIATE_PARTITION_ITERATE(lend, LevelIterator);
```
instead.